packages/server-wallet/src/protocols/channel-closer.ts

```diff
@@ -39,226 +36,155 @@ export class ChannelCloser {
-          await tx.rollback(error);
-          attemptAnotherProtocolStep = false;
+      try {
+        if (!ensureAllAllocationItemsAreExternalDestinations(channel)) {
+          response.queueChannel(channel);
+          return;
         }
-      });
-    }
+
+        if (!(await this.areAllFinalStatesSigned(channel, tx, response))) {
+          response.queueChannel(channel);
+          return;
+        }
+
+        if (!(await this.isChannelDefunded(channel, tx))) {
+          response.queueChannel(channel);
+          return;
+        }
+
+        await this.completeObjective(objective, channel, tx, response);
+      } catch (error) {
+        this.logger.error({error}, 'Error taking a protocol step');
+        await tx.rollback(error);
+      }
+    });
   }
 
-  private async signState(
-    protocolState: ProtocolState,
-    turnNum: number,
+  private async areAllFinalStatesSigned(
+    channel: Channel,
     tx: Transaction,
     response: WalletResponse
-  ): Promise<void> {
-    if (!protocolState.app.supported) {
-      throw new Error('Must have a supported state');
+  ): Promise<boolean> {
+    // I want to sign the final state if:
+    // - I haven't yet signed a final state
+    // - and either
+    //    - there's an existing final state (in which case I double sign)
+    //    - or it's my turn (in which case I craft the final state)
+
+    const {latestSignedByMe, supported, support} = channel;
+    if (channel.hasConclusionProof) return true;
+    if (!latestSignedByMe || !supported || !support.length) return false;
+
+    if (channel.myTurn) {
+      // I am the first to sign a final state
+      if (!supported.isFinal) {
+        await this.signState(channel, supported.turnNum + 1, tx, response);
+        return false;
+      }
+      await this.signState(channel, supported.turnNum, tx, response);
+      return channel.hasConclusionProof;
     }
-    const {myIndex, channelId} = protocolState.app;
-    const channel = await Channel.forId(channelId, tx);
-    const vars: StateVariables = {...protocolState.app.supported, turnNum, isFinal: true};
-    const signedState = await this.store.signState(channel, vars, tx);
-    response.queueState(signedState, myIndex, channelId);
+    return false;
   }
 
-  private async defund(protocolState: ProtocolState, tx: Transaction): Promise<void> {
-    switch (protocolState.app.fundingStrategy) {
+  private async isChannelDefunded(c: Channel, tx: Transaction): Promise<boolean> {
+    const {protocolState: ps} = c;
+
+    switch (ps.fundingStrategy) {
       case 'Direct':
-        await this.withdraw(protocolState, tx);
-        break;
-      case 'Ledger':
-        await this.requestLedgerDefunding(protocolState, tx);
-        break;
-      case 'Fake':
+        if (c.protocolState.directFundingStatus === 'Defunded') {
+          return true;
+        }
+        if (!c.chainServiceRequests.find(csr => csr.request === 'withdraw')?.isValid()) {
```

Comment:
I find `!c.chainServiceRequests.find(csr => csr.request === 'withdraw')?.isValid();` pretty hard to parse.
I think it says:
> There is no valid withdraw chain service request

What is a "valid" chain service request?

Is this clearer?
```suggestion
        const existingWithdrawRequest = c.chainServiceRequests.find(csr => csr.request === 'withdraw')?.isValid()
        if (!existingWithdrawRequest) {
```

Reply:
Good point. This logic is going to get refactored in an upcoming PR as the defunding code will get factored out per https://github.com/statechannels/statechannels/issues/3124. So I am tempted not to address this now.

A valid chain service request is https://github.com/statechannels/statechannels/blob/9ca60fdaf99196348aec8651a89349dd22d294fa/packages/server-wallet/src/models/chain-service-request.ts#L57

We allow 3 retries for submitting a chain service request.

Reply:
`isValid` seems like a confusing name for that concept. `isStale`seems more accurate. Even better, calling its complement`failed/isFailed/hasFailed` seems the most clear.